### ISSUE TITLE
Deadzone (and maybe other calcs?) are a box, not a circle

### ISSUE THREAD

**Issue by evilC** [evilC/UCR#74]
Maybe need circle deadzone etc for XBOX controller users, as the throw of an xbox stick is circular, not square like a flightstick.

**Comment by EternalDahaka** [comment]
Do you just need the calculations?

http://www.third-helix.com/2013/04/12/doing-thumbstick-dead-zones-right.html
This is a good reference for that.  Just use the distance from the center to make a circular deadzone(instead of from the axes which makes a square/axial), and rescale the values so you get the full range outside of the deadzone.
You might need to translate the XInput values to a +-1 range if they aren't already.


It sort of looks like one of the things you're doing with this UCR program is input joystick > mouse output so you might want some other stuff for that.
If you're not already aware of how to do that, you can just use some basic trig for that with something like
angle = arctan(stickY/stickX)

and use the sine and cosine for the output movement
outputVector(cos(angle), sin(angle))

Depending on the mouse functions you may need to invert the sine( I used a crude mouse.Position function which edits the pixel movement, but using an actual move function might be different- the angle alone might be sufficient [I'm not a programmer, so I'm ignorant of most things] )


If you already had the latter information, sorry for the fluff.

**Comment by evilC** [comment]
This is a great source of info, many thanks for bringing it to my attention.

**Comment by evilC** [comment]
This image from that guide is kind of misleading.
![DeadZone image](http://www.third-helix.com/images/2013/04/axial-deadzone.jpg)
The entire red area is not "dead". The vertical part is dead for the X axis, the horizontal part is dead for Y.

UCR's deadzone calcs, whilst "square", already re-scale the space outside the deadzone, so input "ramps up" as described in the "The High-Precision Problem" section. The math for that is [here](https://github.com/evilC/UCR/blob/master/Libraries/StickOps/StickOps.ahk#L45)

The only thing that UCR currently does not do is the circular deadzone.
Currently, all UCR's axis remapping plugins are one axis per plugin, so in order to implement this, I would need to make two-axis versions of these plugins (So it has access to the X and Y values) and then implement new 2D deadzone code.
I am currently holding off writing too many plugins, because the possibility of quite major changes still loom, and I do not want to burden myself with having to maintain too many plugins.

There is nothing, however, stopping anyone from writing such a plugin themselves - UCR would not need any modification itself in order to support this.

**Comment by EternalDahaka** [comment]
True, though the article mentions that the view "snaps" to the cardinals, which is accurate to how it works.

![image](https://cloud.githubusercontent.com/assets/16890044/12638362/be5bd0a0-c559-11e5-96bc-b2e465e71cb0.png)

This however would be a more accurate visual representation.


The code is fine, and is basically what the article advocates.  The only difference would be using the magnitude as the "value" variable, simply gotten by the distance formula.
magntitude = root( [stickX]^2 + [stickY]^2 );

A square deadzone is fine though, depending on what it's used for.  For single axis applications it's the same, but for things like mouse movement, you'd get more intuitive and correct movements with a circular deadzone.
